netket/optimizer/qgt/qgt_jacobian.py

```diff
@@ -0,0 +1,290 @@
+    samples,
+    pdf=None,
+    *,
+    dense: bool,
+    mode: Optional[str] = None,
+    holomorphic: Optional[bool] = None,
+    diag_shift: float = 0.0,
+    diag_scale: float = 0.0,
+    chunk_size: Optional[int] = None,
+    **kwargs,
+) -> Union[QGTJacobianDenseT, QGTJacobianPyTreeT]:
+    """
+    Construct a :class:`QGTJacobianDenseT` or :class:`QGTJacobianPyTreeT`
+    starting from the definition of a variational state.
+
+    This type of the output is determined by the `dense` boolean keyword argument,
+    which is mandatory.
+
+    The `pdf` argument can be used to:
+    - perform a full-summation calculation, where pdf is a probability density
+      that sums to 1.
+    - Similarly, it can be `jnp.ones_like(samples.shape[:-1])/Ns` to return
+      the same result as `pdf = None`.
+    - reweight individual samples a 'la importance sampling, where the pdf
+      should now be the importance reweight factor times `1/Ns`.
+
+    Other arguments are as other jacobian QGT constructors, documented below.
+
+    This constructor is separated as it can be used from packages extending
+    NetKet.
+    """
+    if mode is not None and holomorphic is not None:
+        raise ValueError("Cannot specify both `mode` and `holomorphic`.")
+
+    if mode is None:
+        mode = nkjax.jacobian_default_mode(
+            apply_fun,
+            parameters,
+            model_state,
+            samples,
+            holomorphic=holomorphic,
+        )
+
+    if pdf is not None:
+        if not pdf.shape == samples.shape[:-1]:
+            raise ValueError(
+                "The shape of pdf must match the shape of the samples, "
+                f"instead you provided (pdf.shape={pdf.shape}) != "
+                f"(samples.shape={samples.shape[:-1]})"
+            )
+        if pdf.ndim >= 2:
+            pdf = jax.jit(jax.lax.collapse, static_argnums=(1, 2))(pdf, 0, 2)
+
+    if samples.ndim >= 3:
+        # use jit so that we can do it on global shared array
+        samples = jax.jit(jax.lax.collapse, static_argnums=(1, 2))(samples, 0, 2)
+
+    jacobians = nkjax.jacobian(
+        apply_fun,
+        parameters,
+        samples,
+        model_state,
+        mode=mode,
+        pdf=pdf,
+        chunk_size=chunk_size,
+        dense=dense,
+        center=True,
+        _sqrt_rescale=True,
+    )
+
+    shift, offset = to_shift_offset(diag_shift, diag_scale)
+
+    if offset is not None:
+        ndims = 1 if mode != "complex" else 2
+        jacobians, scale = rescale(jacobians, offset, ndims=ndims)
+    else:
+        scale = None
+
+    pars_struct = jax.tree_util.tree_map(
+        lambda x: jax.ShapeDtypeStruct(x.shape, x.dtype), parameters
+    )
+
+    QGT_T = QGTJacobianDenseT if dense else QGTJacobianPyTreeT
+    return QGT_T(
+        O=jacobians,
+        scale=scale,
+        mode=mode,
+        _params_structure=pars_struct,
+        diag_shift=shift,
+        **kwargs,
+    )
+
+
+@partial_from_kwargs
+def QGTJacobianDense(
```

Comment:
this function here, instead, knows about variational states, and does the relevant unpacking and internally calls the function above.